### PR TITLE
[WIP] Groups

### LINKED ISSUE #11: Button group
Group of buttons, no space in between, rounded corners on the ends, should support v-model so you end up with a checkboxes like interaction

### LINKED ISSUE #12: Item group
Make any arbitrary set of items selectable through v-model. This would be used to make the cards selectable and fully clickable on the cards layout

### LINKED ISSUE #18: Tabs
Most often used in dialogs / modals, but also in the right sidebar of item detail pages

### PR DESCRIPTION
**Pull request by Nitwel**
- [x] Item Group
- [x] Button Group
- [ ] Tab Group
- [ ] Chip Group
- [x] Making to much progress

closes #11 closes #18 closes #12

### CODE REVIEW

File: src/components/groups/v-button-group/v-button-group.vue [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+}
+
+export default createComponent({
+	model: {
+		prop: 'input',
+		event: 'change'
+	},
+	props: {
+		input: {
+			type: Array as () => Button[],
```
Comment (rijkvanzanten) [inline src/components/groups/v-button-group/v-button-group.vue]:
```suggestion
			type: Array as PropType<Button[]>,
```

where PropType is imported from @vue/composition-api

File: src/components/groups/v-item-group/v-item-group.vue [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+		},
+		multiple: {
+			type: Boolean,
+			default: false
+		},
+		...sizeProps
+	},
+	setup(props, { emit }) {
+		const sizeClass = useSizeClass(props);
+		let activeCount = computed(() => {
```
Comment (rijkvanzanten) [inline src/components/groups/v-item-group/v-item-group.vue]:
```suggestion
		const activeCount = computed(() => {
```

File: src/components/groups/v-item-group/v-item-group.vue [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+			let count = 0;
+			props.input.forEach(item => {
+				if (item.active) count++;
+			});
+			return count;
+		});
+
+		return { sizeClass, onClick };
+
+		function onClick(index: any) {
```
Comment (rijkvanzanten) [inline src/components/groups/v-item-group/v-item-group.vue]:
```suggestion
		function onClick(index: number) {
```

File: src/components/groups/v-item-group/v-item-group.vue [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+			const output: Item[] = [];
+
+			props.input.forEach(item => {
+				if (props.mandatory && !props.multiple) {
+					item.active = item[props.index] == index;
+				} else if (!props.mandatory && props.multiple) {
```
Comment (rijkvanzanten) [inline src/components/groups/v-item-group/v-item-group.vue]: Lets be explicit in if statements; prefer

```
if (x === false)
```

over

```
if (!x)
```

File: src/components/groups/v-item-group/v-item-group.vue (hunk 3 above)
Comment (Nitwel) [reply src/components/groups/v-item-group/v-item-group.vue]: This would not make a lot of sense if you use a string as a index.

Comment (rijkvanzanten) [reply src/components/groups/v-item-group/v-item-group.vue]: oh my bad, I thought it was the index of the for loop.

Lets rename it to `key` instead and make it `key: string | number` 🙂

File: src/components/groups/v-button-group/v-button-group.readme.md [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+This is an example how a button configuration could look like.
+
+```js
+const buttons = [
+    { name: 'Home', active: true, icon: 'home'},
+    { name: 'Add', active: false, icon: 'add', iconColor: 'var(--green-600)'},
+    { name: 'Remove', active: false, icon: 'remove', iconColor: 'var(--red-600)' },
+    { name: 'Delete', active: false, icon: 'delete' }
+]
+```
```
Comment (rijkvanzanten) [inline src/components/groups/v-button-group/v-button-group.readme.md]: Is this accurate?

File: src/components/groups/v-item-group/v-item-group.vue (hunk 3 above)
Comment (Nitwel) [reply src/components/groups/v-item-group/v-item-group.vue]: Do you mean the prop? If yes, `key` is a unsuseable name in Vue

File: src/components/groups/v-button-group/v-button-group.readme.md [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+Using the buttons array from above.
+
+```html
+<v-button-group v-model="buttons" index="name"/>
```
Comment (rijkvanzanten) [inline src/components/groups/v-button-group/v-button-group.readme.md]: ah that's why it's named `index`. `key` here would be invalid. 

The table component calls this `item-key`. Maybe we should use that? (Or update the item-key in table to be `index`

**Comment by rijkvanzanten** [comment]
I used to do this for the header buttons:


```js
const buttons = [
    { name: 'Home', active: true, icon: 'home'},
    { name: 'Add', active: false, icon: 'add', iconColor: 'var(--green-600)'},
    { name: 'Remove', active: false, icon: 'remove', iconColor: 'var(--red-600)' },
    { name: 'Delete', active: false, icon: 'delete' }
]
```

and it turned out to be not flexible enough.. For example, what happens if we wanted to add a tooltip to one of these, or show the badge?

We'll have to use the nested markup syntax:

```vue
<v-button-group>
  <v-button />
  <v-button />
  <v-button />
</v-button-group>
```

and go from there

File: src/components/groups/v-button-group/v-button-group.readme.md (hunk 5 above)
Comment (Nitwel) [reply src/components/groups/v-button-group/v-button-group.readme.md]: How should this not be accurate?

Comment (rijkvanzanten) [reply src/components/groups/v-button-group/v-button-group.readme.md]: Asking if this is the current way you're supposed to use button group, or if the readme is outdated 🙂  see comment below regarding that

**Comment by Nitwel** [comment]
> and it turned out to be not flexible enough.. For example, what happens if we wanted to add a tooltip to one of these, or show the badge?

What do you mean by badge? I dont think the purpose of the button group is to show a badge? 😄

**Comment by rijkvanzanten** [comment]
You can very well show a badge with a number on one/each of the buttons in a button group to highlight a substate of the selection. The tooltip is a better example though 😉

**Comment by Nitwel** [comment]
Ah! Just confused myself and mixed up badge with chip. XD
Do you think this is possible with another slot?

**Comment by rijkvanzanten** [comment]
> Do you think this is possible with another slot?

The button group should have a single default slot, and style nested v-button items. That allows us to do things like:

```html
<v-button-group>
  <v-badge value="15">
    <v-button>Click me first</v-button>
  </v-badge>

  <v-button>Click me second</v-button>

  <v-button v-tooltip="'Hi!'">Click third</v-button>
</v-button-group>
```

**Comment by Nitwel** [comment]
> The button group should have a single default slot, and style nested v-button items.

Sorry, but I have 0 clue how to do this. 😅 
How do you know internaly which button goes to which value?

**Comment by rijkvanzanten** [comment]
The selected value would be based on the index. We should be able to read the index of the button child relative to it's siblings based on context.children in Vue 🤔

**Pull request closed without merging**